hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/EnhancerImpl.java

```diff
@@ -654,25 +655,50 @@ else if ( access != null && access.load().value() == AccessType.FIELD ) {
 
 	private static class EnhancerClassFileLocator extends ClassFileLocator.ForClassLoader {
 		private final ConcurrentHashMap<String, Resolution> resolutions = new ConcurrentHashMap<>();
+		private final ConcurrentHashMap<String, LoadingLock> loadingLocks = new ConcurrentHashMap<>();
 
 		/**
 		 * Creates a new class file locator for the given class loader.
 		 *
 		 * @param classLoader The class loader to query which must not be the bootstrap class loader, i.e. {@code null}.
 		 */
-		protected EnhancerClassFileLocator(ClassLoader classLoader) {
-			super( classLoader );
+		protected EnhancerClassFileLocator(final ClassLoader classLoader) {
+			super( Objects.requireNonNull( classLoader ) );
 		}
 
 		@Override
-		public Resolution locate(String className) throws IOException {
-			assert className != null;
+		public Resolution locate(final String className) throws IOException {
 			final Resolution resolution = resolutions.get( className );
 			if ( resolution != null ) {
 				return resolution;
 			}
 			else {
-				return super.locate( className );
+				//N.B. ByteBuddy also implements caching, but has no protections against
+				//a cache stampede as it's not meant to process classes in parallel.
+				//Observations show that it's actually very likely to happen in our case.
+				final LoadingLock newLock = new LoadingLock();
+				final LoadingLock existingLock = loadingLocks.putIfAbsent( className, newLock );
+				final LoadingLock lock = existingLock == null ? newLock : existingLock;
+				synchronized ( lock ) {
+					final Resolution loaded = lock.loaded;
+					if ( loaded != null ) {
+						loadingLocks.remove( className );
+						return loaded;
+					}
```

Comment:
Wouldn't it be better to use a lock instead of synchronized?

```suggestion
				final ReentrantLock lock = loadingLocks.computeIfAbsent( className, k -> new ReentrantLock() );
				lock.lock();
				try {
```

and then `lock.unlock();` in a finally.

Reply:
humm perhaps - you're thinking of Loom?

I was quite happy with my solution of being able to extract the value from the lock, it can work as a value holder as well. But it was late night, perhaps not wise to come with a newly invented scheme :)

I actually wonder why I am needing this complex scheme - it's all to not duplicate the cache that Bytebuddy is already having. Perhaps we're better off by caching it all ourselves and disable the BB cache... I'll try that.